zipkin-server/src/main/java/zipkin2/server/internal/brave/TracingConfiguration.java

```diff
@@ -131,52 +118,76 @@
         Span v2Span = SpanBytesDecoder.PROTO3.decodeOne(encodedSpan);
         spans.add(v2Span);
       }
-      return delegate.spanConsumer().accept(spans);
+      return delegate().spanConsumer().accept(spans);
     }
 
     @Override public CheckResult check() {
-      return delegate.check();
+      return delegate().check();
+    }
+
+    @Override public String toString() {
+      // Avoid using the delegate to avoid eagerly loading the bean during initialization
+      return "StorageComponent";
     }
 
     @Override public void close() {
       // don't close delegate as we didn't open it!
     }
+
+    /** Lazy lookup to avoid proxying */
+    StorageComponent delegate() {
+      StorageComponent result = delegate;
+      if (result != null) return delegate;
+      result = factory.getBean(StorageComponent.class);
```

Comment:
Should this be synchronized i.e. double check?

Reply:
I've noticed a lot of similar code not synchronizing on `factory.getBean` and my guess is that it should be non-'side effect causing' to call this due to a stale field read. In other words it could possibly cost less to occasionally redo factory.getBean vs synchronizing.

@wilkinsona did I reverse-engineer intent right on this pattern? just using you as a spring set of eyes, not because you established this pattern :P

Reply:
Yeah, I think you have. FWIW, when using this pattern I would declare the field `volatile` to eliminate the possibility of a stale read of the field. There's still a potential race between multiple threads, but duplicate calls to `factory.getBean(StorageComponent.class)` should not have any ill effects.

Reply:
thanks for the feedback folks!